File: android-native/ComprehensiveAnalytics.kt
package com.pos54link.app.analytics

import android.content.Context
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import io.sentry.Sentry
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.sql.DriverManager
import java.util.*

// MARK: - Comprehensive Analytics with Platform Integration

class ComprehensiveAnalyticsManager(private val context: Context) {
    
    private val firebaseAnalytics = FirebaseAnalytics.getInstance(context)
    private val lakehouseURL = "https://lakehouse.remittance.app/api/v1/events"
    private val middlewareURL = "https://middleware.remittance.app/api/v1/analytics"
    private val eventQueue = mutableListOf<AnalyticsEvent>()
    private val batchSize = 50
    private val client = OkHttpClient()
    
    companion object {
        @Volatile
        private var instance: ComprehensiveAnalyticsManager? = null
        
        fun getInstance(context: Context): ComprehensiveAnalyticsManager {
            return instance ?: synchronized(this) {
                instance ?: ComprehensiveAnalyticsManager(context).also { instance = it }
            }
        }
    }
    
    init {
        startBatchProcessor()
    }
    
    // MARK: - Event Tracking
    
    fun trackEvent(name: String, parameters: Map<String, Any> = emptyMap()) {
        val event = AnalyticsEvent(
            id = UUID.randomUUID().toString(),
            name = name,
            parameters = parameters,
            timestamp = Date(),
            userId = getCurrentUserId(),
            sessionId = getCurrentSessionId(),
            deviceInfo = getDeviceInfo()
        )
        
        // Firebase Analytics
        val bundle = android.os.Bundle()
        parameters.forEach { (key, value) ->
            when (value) {
                is String -> bundle.putString(key, value)
                is Int -> bundle.putInt(key, value)
                is Long -> bundle.putLong(key, value)
                is Double -> bundle.putDouble(key, value)
                is Boolean -> bundle.putBoolean(key, value)
            }
        }
        firebaseAnalytics.logEvent(name, bundle)
        
        // Add to queue
        synchronized(eventQueue) {
            eventQueue.add(event)
            if (eventQueue.size >= batchSize) {
                flushEvents()
            }
        }
    }
    
    // MARK: - User Acquisition
    
    fun trackUserAcquisition(source: String, medium: String, campaign: String) {
        trackEvent("user_acquisition", mapOf(
            "source" to source,
            "medium" to medium,
            "campaign" to campaign,
            "install_date" to System.currentTimeMillis()
        ))
        
        storeAcquisitionData(source, medium, campaign)
    }
    
    private fun storeAcquisitionData(source: String, medium: String, campaign: String) {
        GlobalScope.launch(Dispatchers.IO) {
            try {
                val connection = DriverManager.getConnection(
                    "jdbc:postgresql://postgres.remittance.app:5432/remittance_analytics",
                    "analytics_user",
                    System.getenv("POSTGRES_PASSWORD") ?: ""
                )
                
                val sql = "INSERT INTO user_acquisition (user_id, source, medium, campaign, created_at) VALUES (?, ?, ?, ?, NOW())"
                val statement = connection.prepareStatement(sql)
                statement.setString(1, getCurrentUserId())
                statement.setString(2, source)
                statement.setString(3, medium)
                statement.setString(4, campaign)
                statement.executeUpdate()
                
                connection.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
    
    // MARK: - Feature Adoption
    
    fun trackFeatureUsage(featureName: String, firstTime: Boolean = false) {
        trackEvent("feature_used", mapOf(
            "feature_name" to featureName,
            "first_time" to firstTime
        ))
    }
    
    // MARK: - Session Tracking
    
    fun startSession() {
        val sessionId = UUID.randomUUID().toString()
        val prefs = context.getSharedPreferences("analytics", Context.MODE_PRIVATE)
        prefs.edit()
            .putString("current_session_id", sessionId)
            .putLong("session_start_time", System.currentTimeMillis())
            .apply()
        
        trackEvent("session_start", mapOf("session_id" to sessionId))
    }
    
    fun endSession() {
        val prefs = context.getSharedPreferences("analytics", Context.MODE_PRIVATE)
        val startTime = prefs.getLong("session_start_time", 0)
        val duration = System.currentTimeMillis() - startTime
        
        trackEvent("session_end", mapOf(
            "session_id" to getCurrentSessionId(),
            "duration" to duration
        ))
    }
    
    // MARK: - Lakehouse Integration
    
    private fun sendToLakehouse(event: String, data: Map<String, Any>) {
        GlobalScope.launch(Dispatchers.IO) {
            try {
                val json = JSONObject().apply {
                    put("event", event)
                    put("data", JSONObject(data))
                    put("timestamp", System.currentTimeMillis())
                    put("user_id", getCurrentUserId())
                }
                
                val body = json.toString().toRequestBody("application/json".toMediaType())
                val request = Request.Builder()
                    .url(lakehouseURL)
                    .post(body)
                    .addHeader("Authorization", "Bearer ${getLakehouseToken()}")
                    .build()
                
                client.newCall(request).execute()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
    
    // MARK: - Batch Processing
    
    private fun startBatchProcessor() {
        GlobalScope.launch(Dispatchers.IO) {
            while (true) {
                delay(60000) // 1 minute
                flushEvents()
            }
        }
    }
    
    private fun flushEvents() {
        val eventsToSend = synchronized(eventQueue) {
            val events = eventQueue.toList()
            eventQueue.clear()
            events
        }
        
        if (eventsToSend.isEmpty()) return
        
        sendToMiddleware(eventsToSend)
    }
    
    private fun sendToMiddleware(events: List<AnalyticsEvent>) {
        GlobalScope.launch(Dispatchers.IO) {
            try {
                val jsonArray = JSONArray()
                events.forEach { event ->
                    jsonArray.put(JSONObject(event.toDictionary()))
                }
                
                val body = jsonArray.toString().toRequestBody("application/json".toMediaType())
                val request = Request.Builder()
                    .url(middlewareURL)
                    .post(body)
                    .build()
                
                client.newCall(request).execute()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
    
    // MARK: - Helper Methods
    
    private fun getCurrentUserId(): String {
        val prefs = context.getSharedPreferences("analytics", Context.MODE_PRIVATE)
        return prefs.getString("user_id", "anonymous") ?: "anonymous"
    }
    
    private fun getCurrentSessionId(): String {
        val prefs = context.getSharedPreferences("analytics", Context.MODE_PRIVATE)
        return prefs.getString("current_session_id", "unknown") ?: "unknown"
    }
    
    private fun getDeviceInfo(): Map<String, Any> {
        return mapOf(
            "model" to android.os.Build.MODEL,
            "os_version" to android.os.Build.VERSION.RELEASE,
            "app_version" to context.packageManager.getPackageInfo(context.packageName, 0).versionName
        )
    }
    
    private fun getLakehouseToken(): String {
        return System.getenv("LAKEHOUSE_TOKEN") ?: ""
    }
}

data class AnalyticsEvent(
    val id: String,
    val name: String,
    val parameters: Map<String, Any>,
    val timestamp: Date,
    val userId: String,
    val sessionId: String,
    val deviceInfo: Map<String, Any>
) {
    fun toDictionary(): Map<String, Any> {
        return mapOf(
            "id" to id,
            "name" to name,
            "parameters" to parameters,
            "timestamp" to timestamp.time,
            "user_id" to userId,
            "session_id" to sessionId,
            "device_info" to deviceInfo
        )
    }
}

// MARK: - A/B Testing

class ABTestingManager(private val context: Context) {
    private val remoteConfig = FirebaseRemoteConfig.getInstance()
    
    fun initialize() {
        remoteConfig.setConfigSettingsAsync(
            com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings.Builder()
                .setMinimumFetchIntervalInSeconds(3600)
                .build()
        )
        
        remoteConfig.setDefaultsAsync(mapOf(
            "onboarding_variant" to "control",
            "button_color" to "#007AFF",
            "pricing_variant" to "monthly"
        ))
        
        remoteConfig.fetchAndActivate()
    }
    
    fun getVariant(experiment: String): String {
        return remoteConfig.getString(experiment)
    }
}

// MARK: - Revenue Tracking with TigerBeetle

class RevenueTrackingManager {
    private val tigerBeetleURL = "https://tigerbeetle.remittance.app/api/v1/revenue"
    private val client = OkHttpClient()
    
    fun trackTransaction(amount: Double, currency: String, type: String) {
        GlobalScope.launch(Dispatchers.IO) {
            try {
                val json = JSONObject().apply {
                    put("id", UUID.randomUUID().toString())
                    put("amount", amount)
                    put("currency", currency)
                    put("type", type)
                    put("timestamp", System.currentTimeMillis())
                }
                
                val body = json.toString().toRequestBody("application/json".toMediaType())
                val request = Request.Builder()
                    .url(tigerBeetleURL)
                    .post(body)
                    .build()
                
                client.newCall(request).execute()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}

File: ios-native/ComprehensiveAnalytics.swift
import Foundation
import FirebaseAnalytics
import PostgresClientKit

// MARK: - Comprehensive Analytics with Lakehouse Integration

class ComprehensiveAnalyticsManager {
    
    static let shared = ComprehensiveAnalyticsManager()
    
    private let lakehouseURL = "https://lakehouse.remittance.app/api/v1/events"
    private let middlewareURL = "https://middleware.remittance.app/api/v1/analytics"
    private var postgresConnection: Connection?
    private var eventQueue: [AnalyticsEvent] = []
    private let batchSize = 50
    
    private init() {
        setupPostgresConnection()
        startBatchProcessor()
    }
    
    // MARK: - Postgres Connection
    
    private func setupPostgresConnection() {
        do {
            var configuration = PostgresClientKit.ConnectionConfiguration()
            configuration.host = "postgres.remittance.app"
            configuration.port = 5432
            configuration.database = "remittance_analytics"
            configuration.user = "analytics_user"
            configuration.credential = .scramSHA256(password: ProcessInfo.processInfo.environment["POSTGRES_PASSWORD"] ?? "")
            configuration.ssl = true
            
            postgresConnection = try PostgresClientKit.Connection(configuration: configuration)
        } catch {
            print("Postgres connection failed: \(error)")
        }
    }
    
    // MARK: - Event Tracking
    
    func trackEvent(_ name: String, parameters: [String: Any] = [:]) {
        let event = AnalyticsEvent(
            id: UUID().uuidString,
            name: name,
            parameters: parameters,
            timestamp: Date(),
            userId: getCurrentUserId(),
            sessionId: getCurrentSessionId(),
            deviceInfo: getDeviceInfo()
        )
        
        // Firebase Analytics
        Analytics.logEvent(name, parameters: parameters)
        
        // Add to queue for batch processing
        eventQueue.append(event)
        
        if eventQueue.count >= batchSize {
            flushEvents()
        }
    }
    
    // MARK: - User Acquisition Tracking
    
    func trackUserAcquisition(source: String, medium: String, campaign: String) {
        trackEvent("user_acquisition", parameters: [
            "source": source,
            "medium": medium,
            "campaign": campaign,
            "install_date": Date().timeIntervalSince1970
        ])
        
        // Store in Postgres
        storeAcquisitionData(source: source, medium: medium, campaign: campaign)
    }
    
    private func storeAcquisitionData(source: String, medium: String, campaign: String) {
        guard let connection = postgresConnection else { return }
        
        let sql = """
        INSERT INTO user_acquisition (user_id, source, medium, campaign, created_at)
        VALUES ($1, $2, $3, $4, NOW())
        """
        
        do {
            let statement = try connection.prepareStatement(text: sql)
            try statement.execute(parameterValues: [
                getCurrentUserId(),
                source,
                medium,
                campaign
            ])
        } catch {
            print("Failed to store acquisition data: \(error)")
        }
    }
    
    // MARK: - Onboarding Tracking
    
    func trackOnboardingStep(_ step: Int, completed: Bool) {
        trackEvent("onboarding_step", parameters: [
            "step": step,
            "completed": completed,
            "completion_time": Date().timeIntervalSince1970
        ])
        
        if completed && step == 5 { // Final step
            trackEvent("onboarding_completed", parameters: [
                "total_time": calculateOnboardingDuration()
            ])
        }
    }
    
    private func calculateOnboardingDuration() -> TimeInterval {
        // Calculate from first step to completion
        return 0.0 // Implement based on stored timestamps
    }
    
    // MARK: - Feature Adoption Tracking
    
    func trackFeatureUsage(_ featureName: String, firstTime: Bool = false) {
        trackEvent("feature_used", parameters: [
            "feature_name": featureName,
            "first_time": firstTime,
            "usage_count": getFeatureUsageCount(featureName) + 1
        ])
        
        // Update feature adoption in Postgres
        updateFeatureAdoption(featureName: featureName)
    }
    
    private func updateFeatureAdoption(featureName: String) {
        guard let connection = postgresConnection else { return }
        
        let sql = """
        INSERT INTO feature_adoption (user_id, feature_name, first_used_at, usage_count)
        VALUES ($1, $2, NOW(), 1)
        ON CONFLICT (user_id, feature_name)
        DO UPDATE SET usage_count = feature_adoption.usage_count + 1, last_used_at = NOW()
        """
        
        do {
            let statement = try connection.prepareStatement(text: sql)
            try statement.execute(parameterValues: [getCurrentUserId(), featureName])
        } catch {
            print("Failed to update feature adoption: \(error)")
        }
    }
    
    private func getFeatureUsageCount(_ featureName: String) -> Int {
        // Query from Postgres
        return 0
    }
    
    // MARK: - Retention Metrics
    
    func trackRetention(day: Int) {
        trackEvent("retention_check", parameters: [
            "day": day,
            "retained": true
        ])
        
        // Store in Lakehouse for cohort analysis
        sendToLakehouse(event: "retention", data: [
            "user_id": getCurrentUserId(),
            "day": day,
            "cohort_date": getUserCohortDate()
        ])
    }
    
    private func getUserCohortDate() -> String {
        // Get user's install date
        return "2025-01-01"
    }
    
    // MARK: - Session Tracking
    
    func startSession() {
        let sessionId = UUID().uuidString
        UserDefaults.standard.set(sessionId, forKey: "current_session_id")
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: "session_start_time")
        
        trackEvent("session_start", parameters: [
            "session_id": sessionId
        ])
    }
    
    func endSession() {
        let duration = Date().timeIntervalSince1970 - (UserDefaults.standard.double(forKey: "session_start_time"))
        
        trackEvent("session_end", parameters: [
            "session_id": getCurrentSessionId(),
            "duration": duration
        ])
        
        // Store session data in Postgres
        storeSessionData(duration: duration)
    }
    
    private func storeSessionData(duration: TimeInterval) {
        guard let connection = postgresConnection else { return }
        
        let sql = """
        INSERT INTO user_sessions (user_id, session_id, duration, created_at)
        VALUES ($1, $2, $3, NOW())
        """
        
        do {
            let statement = try connection.prepareStatement(text: sql)
            try statement.execute(parameterValues: [
                getCurrentUserId(),
                getCurrentSessionId(),
                Int(duration)
            ])
        } catch {
            print("Failed to store session data: \(error)")
        }
    }
    
    // MARK: - Screen View Tracking
    
    func trackScreenView(_ screenName: String) {
        trackEvent("screen_view", parameters: [
            "screen_name": screenName,
            "previous_screen": getPreviousScreen()
        ])
        
        UserDefaults.standard.set(screenName, forKey: "previous_screen")
    }
    
    private func getPreviousScreen() -> String {
        return UserDefaults.standard.string(forKey: "previous_screen") ?? "none"
    }
    
    // MARK: - Button Click Tracking
    
    func trackButtonClick(_ buttonName: String, screen: String) {
        trackEvent("button_click", parameters: [
            "button_name": buttonName,
            "screen": screen
        ])
    }
    
    // MARK: - Error Tracking
    
    func trackError(_ error: Error, context: String) {
        trackEvent("error_occurred", parameters: [
            "error_message": error.localizedDescription,
            "context": context,
            "error_code": (error as NSError).code
        ])
        
        // Send to Lakehouse for error analysis
        sendToLakehouse(event: "error", data: [
            "error": error.localizedDescription,
            "context": context,
            "stack_trace": Thread.callStackSymbols.joined(separator: "\n")
        ])
    }
    
    // MARK: - Crash-Free Rate Tracking
    
    func trackCrashFreeSession() {
        trackEvent("crash_free_session", parameters: [
            "session_id": getCurrentSessionId()
        ])
    }
    
    // MARK: - Lakehouse Integration
    
    private func sendToLakehouse(event: String, data: [String: Any]) {
        guard let url = URL(string: lakehouseURL) else { return }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(getLakehouseToken())", forHTTPHeaderField: "Authorization")
        
        let payload: [String: Any] = [
            "event": event,
            "data": data,
            "timestamp": Date().timeIntervalSince1970,
            "user_id": getCurrentUserId()
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                print("Lakehouse upload failed: \(error)")
            }
        }.resume()
    }
    
    // MARK: - Batch Processing
    
    private func startBatchProcessor() {
        Timer.scheduledTimer(withTimeInterval: 60, repeats: true) { [weak self] _ in
            self?.flushEvents()
        }
    }
    
    private func flushEvents() {
        guard !eventQueue.isEmpty else { return }
        
        let eventsToSend = eventQueue
        eventQueue.removeAll()
        
        // Send to Middleware
        sendToMiddleware(events: eventsToSend)
        
        // Store in Postgres
        storeEventsInPostgres(events: eventsToSend)
    }
    
    private func sendToMiddleware(events: [AnalyticsEvent]) {
        guard let url = URL(string: middlewareURL) else { return }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let payload = events.map { $0.toDictionary() }
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)
        
        URLSession.shared.dataTask(with: request).resume()
    }
    
    private func storeEventsInPostgres(events: [AnalyticsEvent]) {
        guard let connection = postgresConnection else { return }
        
        let sql = """
        INSERT INTO analytics_events (id, name, parameters, user_id, session_id, created_at)
        VALUES ($1, $2, $3, $4, $5, $6)
        """
        
        do {
            for event in events {
                let statement = try connection.prepareStatement(text: sql)
                let parametersJSON = try JSONSerialization.data(withJSONObject: event.parameters)
                
                try statement.execute(parameterValues: [
                    event.id,
                    event.name,
                    String(data: parametersJSON, encoding: .utf8) ?? "{}",
                    event.userId,
                    event.sessionId,
                    event.timestamp
                ])
            }
        } catch {
            print("Failed to store events in Postgres: \(error)")
        }
    }
    
    // MARK: - Helper Methods
    
    private func getCurrentUserId() -> String {
        return UserDefaults.standard.string(forKey: "user_id") ?? "anonymous"
    }
    
    private func getCurrentSessionId() -> String {
        return UserDefaults.standard.string(forKey: "current_session_id") ?? "unknown"
    }
    
    private func getDeviceInfo() -> [String: Any] {
        return [
            "model": UIDevice.current.model,
            "os_version": UIDevice.current.systemVersion,
            "app_version": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown"
        ]
    }
    
    private func getLakehouseToken() -> String {
        return ProcessInfo.processInfo.environment["LAKEHOUSE_TOKEN"] ?? ""
    }
}

// MARK: - Analytics Event Model

struct AnalyticsEvent {
    let id: String
    let name: String
    let parameters: [String: Any]
    let timestamp: Date
    let userId: String
    let sessionId: String
    let deviceInfo: [String: Any]
    
    func toDictionary() -> [String: Any] {
        return [
            "id": id,
            "name": name,
            "parameters": parameters,
            "timestamp": timestamp.timeIntervalSince1970,
            "user_id": userId,
            "session_id": sessionId,
            "device_info": deviceInfo
        ]
    }
}
